### DISCUSSION TITLE
Unable to use Cloudflare's Durable Objects with Remix when using Cloudflare Pages

### DISCUSSION THREAD

**Discussion by davidbarratt** [remix-run/remix#3066] [Proposals]
Currently the only way to use [Durable Objects](https://developers.cloudflare.com/workers/learning/using-durable-objects/) when using [Cloudflare Pages](https://pages.cloudflare.com/) is by having a completely separate Worker and communicating with it from Remix.

This is because there is no way to `export` the Durable Object `class` from a [Pages Function](https://developers.cloudflare.com/pages/platform/functions/).

However, this **can** be done in [Advanced Mode](https://developers.cloudflare.com/pages/platform/functions/#advanced-mode).

Since Remix takes over **all** paths anyways (i.e. `functions/[[path]].js`), why not use `_worker.js` instead? This would allow the user to modify it at the top level which is more what I would have expected from using Pages with Remix anyways.

There is a (somewhat hacky) workaround.

```ts
// server.ts
import { createPagesFunctionHandler } from "@remix-run/cloudflare-pages";
import * as build from "@remix-run/dev/server-build";

const handleRequest = createPagesFunctionHandler({
  build,
  mode: process.env.NODE_ENV,
  getLoadContext: (context) => context.env,
});

interface Env {
  LEAD: DurableObjectNamespace
}

const handler: ExportedHandler<Env> = {
  async fetch(request, env, ctx) {
    return handleRequest({
      request: new Request(request),
      env,
      waitUntil: ctx.waitUntil,
      params: {},
      data: undefined,
      next: () => {
        throw new Error('next() called in Worker')
      }
    });
  }
};

export default handler;

// Durable Object
export { Lead } from '~/do/lead';
```

```js
// remix.config.js
/**
 * @type {import('@remix-run/dev').AppConfig}
 */
module.exports = {
  serverBuildTarget: "cloudflare-pages",
  server: "./server.ts",
  devServerBroadcastDelay: 1000,
  ignoredRouteFiles: [".*"],
  // appDirectory: "app",
  // assetsBuildDirectory: "public/build",
  serverBuildPath: "public/_worker.js",
  // publicPath: "/build/",
};
```
This works, but it's not the cleanest implementation.

I think Remix should provide a handler for Pages Functions in "Advanced" mode rather than the simple mode. This would allow you to do more advanced things like working with Durable Objects, etc.

**Comment by tom-sherman** [comment]
You can define your own worker in the functions/ directory, I think it needs to be in a subfolder though.

Example: https://github.com/tom-sherman/remix-cloudflare-pages-demo/pull/1

**Reply by rodbs** [reply]
@tom-sherman  I don't see the subfolder you mention.

**Reply by rodbs** [reply]
@davidbarratt I'm suffering the same thing. Is this _worker.js the only workaround to make it work? Still not solution from Remix?

**Reply by tom-sherman** [reply]
Inside functions/api

**Comment by aaronadamsCA** [comment]
Another reason for Remix to use `_worker.js` instead of `functions/[[path]].js` would be improved flexibility in the following scenarios:

- Compatibility with drag-and-drop deployments, [which don't currently support a `functions/` directory](https://developers.cloudflare.com/pages/platform/direct-upload/#using-functions).
- ~~Better compatibility with the GitHub Action for Cloudflare Pages, which [doesn't currently support functions from a monorepo](https://github.com/cloudflare/pages-action/issues/1).~~
  - This is especially important because [direct upload is currently the only method to deploy from a monorepo to Cloudflare Pages](https://developers.cloudflare.com/pages/platform/known-issues/#builds-and-deployment).

Right now, it seems our only option to deploy a Remix app from a monorepo to Cloudflare Pages using GitHub Actions is to write our own JS workflow. Granted, improvements to Cloudflare Pages would resolve this; but so would Remix using `_worker.js`, FWIW.

**Reply by ChristianJacobsen** [reply]
I'm facing the exact same issues as you are, @aaronadamsCA. Currently, my only option is to skip using the GH Action for CP and manually upload both functions and public dirs with Wrangler. It's less than ideal, and I sorely wish the Remix team would devote more attention to CF Pages.

**Reply by aaronadamsCA** [reply]
@ChristianJacobsen I actually filed https://github.com/cloudflare/pages-action/pull/50 3 weeks ago to add monorepo support, but there's been no movement so far. I just bumped it.

In the meantime, you're free to fork https://github.com/nisheretail/cloudflare-pages-action/tree/add-working-directory which contains the fix.

**Reply by ChristianJacobsen** [reply]
Thanks, @aaronadamsCA! You're a gem!

**Reply by aaronadamsCA** [reply]
NP. That PR is merged, now we just need https://github.com/cloudflare/pages-action/pull/56 merged and it should be functionally equal to the fork.

**Reply by ChristianJacobsen** [reply]
Perfect! That and a new release, and we should be good to go 🎉 Thanks for the amazing work!

**Reply by hiendaovinh** [reply]
Hi, do we have any updates on this?

**Reply by praneybehl** [reply]
Yeah looking for the same, did it ever make it out?

**Reply by aaronadamsCA** [reply]
If we're talking about the OP issue, no, I am pretty sure `functions/[[path]].ts` remains the only way to use Remix with Cloudflare Pages.

If we're talking about the specific fixes to the Cloudflare Pages action to support monorepos, yes, those changes went out long ago—although if I were implementing from scratch, I would strongly suggest using the Wrangler action instead, since the Pages action is of really poor quality.

**Comment by candidia** [comment]
Has anything changed, especially with vite now?

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
